Corona-Warn-App/src/main/java/de/rki/coronawarnapp/task/TaskController.kt

```diff
@@ -0,0 +1,215 @@
+        }
+        Timber.tag(TAG).i("Task submitted: %s", request)
+        taskQueue.offer(request)
+    }
+
+    suspend fun cancel(requestId: UUID) = internalTaskData.updateSafely {
+        val taskState = values.singleOrNull {
+            it.request.id == requestId
+        } ?: throw IllegalArgumentException("No task found for request ID $requestId")
+
+        Timber.tag(TAG).w("Manually canceling %s", taskState)
+        taskState.deferred.cancel(cause = TaskCancelationException())
+    }
+
+    private suspend fun initTaskData(newRequest: TaskRequest) {
+        Timber.tag(TAG).d("Processing new task request: %s", newRequest)
+        val taskFactory = taskFactories[newRequest.type.java]
+        requireNotNull(taskFactory) { "No factory available for $newRequest" }
+
+        Timber.tag(TAG).v("Initiating task data for request: %s", newRequest)
+        val taskConfig = taskFactory.config
+        val task = taskFactory.taskProvider()
+
+        val deferred = taskScope.async(
+            start = CoroutineStart.LAZY
+        ) { task.run(newRequest.arguments) }
+
+        val activeTask = InternalTaskState(
+            request = newRequest,
+            createdAt = timeStamper.nowUTC,
+            config = taskConfig,
+            task = task,
+            deferred = deferred
+        )
+
+        internalTaskData.updateSafely {
+            val existingRequest = values.singleOrNull { it.request.id == newRequest.id }
+            if (existingRequest == null) {
+                Timber.tag(TAG).d("Added new pending task: %s", activeTask)
+                put(activeTask.id, activeTask)
+            } else {
+                Timber.tag(TAG).w(
+                    "TaskRequest was already used. Existing: %s\nNew request: %s",
+                    existingRequest, newRequest
+                )
+            }
+        }
+        taskScope.launch { processMap() }
+    }
+
+    private suspend fun processMap() = internalTaskData.updateSafely {
+        Timber.tag(TAG).d("Processing task data (count=%d)", size)
+        Timber.tag(TAG).v("Tasks before processing: %s", this.values)
+
+        // Procress all unprocessed finished tasks
+        values.toList()
+            .filter { it.deferred.isCompleted && it.state != TaskState.State.FINISHED }
+            .forEach { data ->
+                val error = data.deferred.getCompletionExceptionOrNull()
+                val result = if (error == null) {
+                    data.deferred.getCompleted()
+                } else {
+                    Timber.tag(TAG).e(error, "Task failed: %s", data)
+                    null
+                }
+
+                this[data.id] = data.copy(
+                    completedAt = timeStamper.nowUTC,
+                    result = result,
+                    error = error
+                )
+                Timber.tag(TAG).i("Task is now FINISHED: %s", data)
+            }
+
+        // Start new tasks
+        values.toList()
```

Comment:
Could be moved to a dedicated function

Reply:
I don't like the idea of extracting methods that are called only from a single location. This method is not too complex and sufficiently documented in my opinion.

Reply:
I don't like the idea of extracting methods that are called only from a single location. This method is not too complex and sufficiently documented in my opinion.